component/capacity.libsonnet

```diff
@@ -21,31 +22,62 @@ local predict(indicator, range='1d', resolution='5m', predict='3*24*60*60') =
   'predict_linear(avg_over_time(%(indicator)s[%(range)s:%(resolution)s])[%(range)s:%(resolution)s], %(predict)s)' %
   { indicator: indicator, range: range, resolution: resolution, predict: predict };
 
-local filterWorkerNodes(metric, workerRole='app', nodeLabel='node') =
-  '%(metric)s * on(%(nodeLabel)s) group_left label_replace(kube_node_role{role="%(workerRole)s"}, "%(nodeLabel)s", "$1", "node", "(.+)")' %
-  { metric: metric, nodeLabel: nodeLabel, workerRole: workerRole };
 
-local maxPerNode(resource) = 'max((%s) * on(node) group_left kube_node_role{role="app"})' % resource;
+local addNodeLabels(metric) =
+  if std.length(byLabels) > 0 then
+    '%(metric)s * on(node) group_left(%(labelList)s) kube_node_labels' %
+    { metric: metric, labelList: std.join(', ', byLabels) }
+  else
+    metric;
+
+local renameNodeLabel(expression, nodeLabel) =
+  if nodeLabel == 'node' then
+    expression
+  else
+    'label_replace(%(expression)s, "node", "$1", "%(nodeLabel)s", "(.+)")' %
+    { expression: expression, nodeLabel: nodeLabel };
 
-local resourceCapacity(resource) = 'sum(%s)' % filterWorkerNodes('kube_node_status_capacity{resource="%s"}' % resource);
-local resourceAllocatable(resource) = 'sum(%s)' % filterWorkerNodes('kube_node_status_allocatable{resource="%s"}' % resource);
-local resourceRequests(resource) = 'sum(%s)' % filterWorkerNodes('kube_pod_resource_request{resource="%s"}' % resource);
+local filterWorkerNodes(metric, workerRole='app', nodeLabel='node') =
+  addNodeLabels(
+    '%(metric)s * on(node) group_left kube_node_role{role="%(workerRole)s"}' %
+    { metric: renameNodeLabel(metric, nodeLabel), nodeLabel: nodeLabel, workerRole: workerRole }
+  );
+
+local aggregate(expression, aggregator='sum') =
```

Comment:
very personal opinion™️: If defining a function i try to get all parameters into the signature and try to not access global state.

Specifically the `byLabels` confused me where it is coming from.

Reply:
In this particular case, I disagree: We have about 3 levels of functions calling each other, all in the scope of this file. Following the parameter through all the function calls is, in my personal opinion ™️ , harder to understand than finding the one global parameter.